src/util/server.ml

```diff
@@ -98,26 +110,57 @@ let bind () =
     Sys.remove path;
     Some (Unix.input_of_descr conn), Some (Unix.output_of_descr conn))
 
-let start file do_analyze =
+let start file preprocess_and_merge do_analyze =
   let input, output = bind () in
   GobConfig.set_bool "incremental.save" true;
-  serve (make file do_analyze ?input ?output)
-
-let analyze ?(reset=false) ({ file; do_analyze; _ }: t)=
+  serve (make file preprocess_and_merge do_analyze ?input ?output)
+
+let reparse (s: t) =
+  if GobConfig.get_bool "server.reparse" then (
+    Goblintutil.create_temp_dir ();
+    Fun.protect ~finally:Goblintutil.remove_temp_dir s.preprocess_and_merge, true)
+  else s.file, false
+
+(* Only called when the file has not been reparsed, so we can skip the expensive CFG comparison. *)
+let virtual_changes file =
+  let changes = CompareCIL.empty_change_info () in
+  let reanalyze = GobConfig.get_string_list "incremental.force-reanalyze.funs" in
+  Cil.iterGlobals file (
+    function
+    | GFun (fundec, _) as global ->
+      if List.mem fundec.svar.vname reanalyze then
+        changes.changed <- { old = global; current = global; unchangedHeader = true; diff = None } :: changes.changed
+      else
+        changes.unchanged <- global :: changes.unchanged
+    | global -> changes.unchanged <- global :: changes.unchanged);
+  changes
+
+let increment_data (s: t) file reparsed = match !Serialize.server_solver_data with
+  | Some solver_data when reparsed ->
+    let _, changes = VersionLookup.updateMap s.file file s.version_map in
+    let old_data = Some { Analyses.cil_file = s.file; solver_data } in
+    s.max_ids <- UpdateCil.update_ids s.file s.max_ids file s.version_map changes;
+    { Analyses.changes; old_data; new_file = file }, false
+  | Some solver_data ->
+    let changes = virtual_changes file in
+    let old_data = Some { Analyses.cil_file = file; solver_data } in
+    { Analyses.changes; old_data; new_file = file }, false
+  | _ -> Analyses.empty_increment_data file, true
```

Comment:
Isn't this duplicating some logic from elsewhere in simplified form? Because these things look very specific to incremental analysis. It would be more error-prone if this logic isn't duplicated in different places because they could easily go out of sync: it's easy to forget to fix things here in server mode if modifying the general incremental logic.